### ISSUE TITLE
allow updating all fields in upsert without re-declaring

### ISSUE THREAD

**Issue by steebchen** [steebchen/prisma-client-go#489]
Creating & Updating the same fields is very repetitive. It would be great to be able to simplify this.

```go
_, err := r.DB.User.UpsertOne(
	db.User.Email.Equals("mock"),
).Create(
	db.User.ID.Set("user_1_mock"),
	db.User.Email.Set("mock"),
	db.User.PasswordHash.Set(r.PW.Hash("mock")),
	db.User.Name.Set("john"),
).Update(
	db.User.ID.Set("user_1_mock"),
	db.User.Email.Set("mock"),
	db.User.PasswordHash.Set(r.PW.Hash("mock")),
	db.User.Name.Set("john"),
).Exec(ctx)
if err != nil {
	return err
}
```

js equivalent:

```js
const data = { id: "user_1_mock", ... }
prisma.user.upsert({
    where: { ... },
    data: data,
    update: data,
})
```

**Comment by steebchen** [comment]
One suggestion: `UpdateAll()`?

**Comment by matthewmueller** [comment]
Upsert?

**Comment by steebchen** [comment]
This IS upsert – it's really just about minimising code if you want to update and create/update all fields in both cases.

**Comment by ynden** [comment]
@steebchen 
So what should it look like in Go? Is something like this valid:
```go
_, err := r.DB.User.UpsertOne(
	db.User.Email.Equals("mock"),
	db.User.ID.Set("user_1_mock"),
	db.User.Email.Set("mock"),
	db.User.PasswordHash.Set(r.PW.Hash("mock")),
	db.User.Name.Set("john"),
).Exec(ctx)
if err != nil {
	return err
}
```

With the function definition as:
```go
UpsertOne(ModelEqualsUniqueWhereParam, ModelWithPrismaPropertySetParam, optional ...ModelSetParam)
```
1. ModelEqualsUniqueWhereParam: Is the where clause
2. ModelWithPrismaPropertySetParam: The mandatory properties (not null)
3. ModelSetParam: The optional properties (that can be null)

This will call the `Create()` and `Update()` functions behind the scenes, so we reduce verbosity.    
    
Otherwise we can also have something with `Data()`:
```go
_, err := r.DB.User.UpsertOne(
	db.User.Email.Equals("mock"),
).Data(
	db.User.ID.Set("user_1_mock"),
	db.User.Email.Set("mock"),
	db.User.PasswordHash.Set(r.PW.Hash("mock")),
	db.User.Name.Set("john"),
).Exec(ctx)
if err != nil {
	return err
}
```

`Data()` will have the `2.` & `3.` parameters from above.

**Comment by steebchen** [comment]
Yeah that sounds good; I'd probably add a new method so that it's not breaking and people can still use the existing methods. I prefer doing something like .UpsertOne().CreateAndUpdate() so that the syntax is the same and you can use CreateAndUpdate() instead of .Create() and .Update().

**Comment by steebchen** [comment]
Thanks for looking into this.

Hm I'm not sure about the MongoDB error; if you have some code already maybe you can push it and I can take a look. I guess maybe it doesn't accept ID in the update params?

**Comment by ynden** [comment]
Yes, you can pull the code from here: https://github.com/ynden/prisma-client-go/tree/feature/upsert-with-id
Run the test setup, and generation, then run the `upsert` tests using:
```bash
cd test
go test -race ./features/upsert
```

**Comment by ynden** [comment]
I see you pulled the `ynden:feature/upsert-with-data` branch, I was actually referring to `ynden:feature/upsert-with-id`, so you can test locally.
The `ynden:feature/upsert-with-data` branch isn't ready yet.
